packages/core/src/types.ts

```diff
@@ -159,11 +159,17 @@ export function indexStructureDefinition(structureDefinition: StructureDefinitio
 
   const elements = structureDefinition.snapshot?.element;
   if (elements) {
-    // First pass, build types
-    elements.forEach((element) => indexType(structureDefinition, element));
+    // @TODO(ThatOneBro 29 Aug 2023): any reason why this can't be done in one loop?
```

Comment:
Good question, and something we should discuss as a team.  There are currently a diversity of opinions about "short functional programming" style vs "fast C-like" style.  I understand there are pros/cons to both.

In this particular case, `indexStructureDefinition` is a not on any critical performance paths, so it is unlikely to have any measurable difference.

But I like the idea that we're aligned as a team, and we could write up a doc that describes our philosophy, and even better, codify it in linter rules.

Reply:
I'm pretty flexible when it comes to how code is written in general, but I do thing `.forEach` is something we should be wary of, in general. I think it becomes easy to forget about potential perf hits especially when it doesn't seem like it may not impact the hot paths, but I think this kind of awareness of when something is okay and when it's not is something that must be fostered in a culture.

For objects you pay most of the cost in the accessing of fields itself, so the performance impact is not as great, but the main thing to be aware of is that there is a non-zero amount of overhead, especially when it comes to `.forEach`. Actually the overhead is quite high for non-objects / primitive values for `.forEach`:

<img width="803" alt="Screen Shot 2023-09-01 at 2 26 30 PM" src="https://github.com/medplum/medplum/assets/1592008/4a708a1d-d97f-4cca-85d7-110384ff8cf5">

Over 30x slower for primitive values, insane!

I agree though, this is definitely a topic we should definitely discuss a group and come up with a proper style guide to follow. I think there is definitely a balance to strike and I think it's easy to get nickel'd-and-dime'd on both sides of the spectrum of performance and readability / complexity.